extensions/blocks/amazon/edit.js

```diff
@@ -1,23 +1,248 @@
 /**
  * External dependencies
  */
+import tinycolor from 'tinycolor2';
 import { __ } from '@wordpress/i18n';
-import { Component } from '@wordpress/element';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	BlockIcon,
+	ContrastChecker,
+	InspectorControls,
+	PanelColorSettings,
+} from '@wordpress/block-editor';
+import {
+	Button,
+	FormTokenField,
+	Notice,
+	PanelBody,
+	Placeholder,
+	ToggleControl,
+} from '@wordpress/components';
+import { useState } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
+import icon from './icon';
+import data from './dummy-data';
 import './editor.scss';
 
-class AmazonEdit extends Component {
-	/**
-	 * Write the block editor UI.
-	 *
-	 * @returns {object} The UI displayed when user edits this block.
-	 */
-	render() {
-		return <p>{ __( 'Coming soon', 'jetpack' ) }</p>;
-	}
-}
+export default function AmazonEdit( {
+	attributes: {
+		backgroundColor,
+		textColor,
+		buttonAndLinkColor,
+		asin,
+		showImage,
+		showTitle,
+		showSeller,
+		showPrice,
+		showPurchaseButton,
+	},
+	className,
+	setAttributes,
+} ) {
+	const notice = false; // TODO
+
+	const [ suggestions, setSuggestions ] = useState( [] );
+	const onInputChange = () => {
+		// TODO get suggestions from API
+		// It would be great if we didn't have to embed the ASIN like this but I think that
+		// requires changes to core
+		setSuggestions(
+			data.products.map( dataProduct => `${ dataProduct.title } (ASIN:${ dataProduct.asin })` )
+		);
+	};
+
+	const idRegex = /^(\d+)$|\(ASIN:(.+)\)$/;
+	const onChange = selectedProducts => {
+		const selectedIds = selectedProducts.map( selectedProduct => {
+			const parsed = idRegex.exec( selectedProduct );
+			const selectedId = parsed[ 1 ] || parsed[ 2 ];
+			return data.products.filter( filteredProduct => filteredProduct.asin === selectedId );
+		} );
+		setAttributes( { asin: selectedIds[ 0 ][ 0 ].asin } );
+	};
+
+	const blockPlaceholder = (
+		<Placeholder
+			label={ __( 'Amazon', 'jetpack' ) }
+			instructions={ __( 'Search by entering an Amazon product name or ID below.', 'jetpack' ) }
+			icon={ <BlockIcon icon={ icon } /> }
+			notices={
+				notice && (
+					<Notice status="error" isDismissible={ false }>
+						{ notice }
+					</Notice>
+				)
+			}
+		>
+			<form>
+				<FormTokenField
+					value={ asin }
+					suggestions={ suggestions }
+					onInputChange={ onInputChange }
+					maxSuggestions={ 10 }
+					label={ __( 'Products', 'jetpack' ) }
+					onChange={ onChange }
+				/>
+				<Button isSecondary isLarge type="submit">
+					{ __( 'Preview', 'jetpack' ) }
+				</Button>
+			</form>
+		</Placeholder>
+	);
+
+	const inspectorControls = (
+		<InspectorControls>
+			{ asin && (
+				<>
+					<PanelBody title={ __( 'Promotion Settings', 'jetpack' ) }>
+						<ToggleControl
```

Comment:
Maybe we should extract this so we don't have to repeat each `ToggleControl` here?

Reply:
I thought about that, but I don't think the abstraction would be helpful. There's not a lot of repeated code.